src/Uno.UI.Composition/Composition/ShapeVisual.skia.cs

```diff
@@ -47,11 +49,14 @@ internal override void Render(in DrawingSession parentSession)
 	/// <inheritdoc />
 	internal override void Draw(in DrawingSession session)
 	{
+		var canvas = session.Canvas;
 		if (ViewBox is { } viewBox)
 		{
-			session.Canvas.ClipRect(viewBox.GetRect(), antialias: true);
+			canvas.ClipRect(viewBox.GetRect(), antialias: true);
 		}
 
+		TotalMatrix = canvas.TotalMatrix;
```

Comment:
Is it possible that we don't read `TotalMatrix` from the canvas, and update it right away when Visual properties change?

That way, we will not even need to calculate transforms on every render frame. So, while rendering we will be just calling `canvas.SetMatrix(TotalMatrix)` where `TotalMatrix` is updated as the Visual properties are being updated. This has to be carefully done though as changes will need to happen recursively to all descendants, and it needs proper handling when a new child is being added.

Reply:
I don't think that's very worthwhile. You'll need to change the matrix on every element while drawing anyway, so you wouldn't gain much by precalculating the matrix for each element. You'd basically be setting the matrix to a new value instead of concatenating the preexisting matrix with the current element's transformation matrix. It shouldn't matter very much I think.

Reply:
It matters from a code clarity/correctness point of view. The idea is to ensure that `TotalMatrix` always has the up-to-date value even if render didn't happen.

Performance-wise, it's unclear to me if it will be slower, faster, or *almost* the same. **Maybe** it won't matter that much and the difference will be too small to be noticable.

Reply:
I've tried this, and it's surprisingly working very well. We should not consider the `TotalMatrix` as the single source of truth of where the element visually actually is (along with the clipping).